containers/DvmState.ts

```diff
@@ -0,0 +1,140 @@
+import { createContainer } from "unstated-next";
+import { useState, useEffect } from "react";
+import { utils } from "ethers";
+
+import { useQuery } from "@apollo/client";
+import { PRICE_REQUESTS } from "../apollo/uma/queries";
+
+import Connection from "./Connection";
+import DvmContract from "./DvmContract";
+import EmpState from "./EmpState";
+
+const { formatUnits: fromWei, parseBytes32String: hexToUtf8 } = utils;
+
+interface ContractState {
+  hasEmpPrice: boolean | null;
+  resolvedPrice: number | null;
+}
+
+const initState = {
+  hasEmpPrice: null,
+  resolvedPrice: null,
+};
+
+// Taken from Voting.sol implementation.
+enum PRICE_REQUEST_STATUSES {
+  NOT_REQUESTED, // Was never requested.
+  ACTIVE, // Is being voted on in the current round.
+  RESOLVED, // Was resolved in a previous round.
+  FUTURE, // Is scheduled to be voted on in a future round.
+}
+
+interface PriceRequestQuery {
+  identifier: {
+    id: string;
+  };
+  time: string;
+}
+
+const useContractState = () => {
+  const { block$ } = Connection.useContainer();
+  const { contract: dvm } = DvmContract.useContainer();
+  const { empState } = EmpState.useContainer();
+  const { priceIdentifier, expirationTimestamp } = empState;
+
+  const [state, setState] = useState<ContractState>(initState);
+
+  // Because apollo caches results of queries, we will poll/refresh this query periodically.
+  // We set the poll interval to a very slow 5 seconds for now since the position states
+  // are not expected to change much.
+  // Source: https://www.apollographql.com/docs/react/data/queries/#polling
+  const { loading, error, data } = useQuery(PRICE_REQUESTS, {
+    context: { clientName: "UMA" },
+    pollInterval: 5000,
+  });
+
+  // get state from EMP
+  const queryState = async () => {
+    if (
+      dvm !== null &&
+      priceIdentifier !== null &&
+      expirationTimestamp !== null
+    ) {
+      const res = await Promise.all([
+        dvm.getPriceRequestStatuses([
+          {
+            identifier: priceIdentifier.toString(),
+            time: expirationTimestamp.toNumber(),
+          },
+        ]),
+      ]);
+
+      const hasPrice = res[0][0].status === PRICE_REQUEST_STATUSES.RESOLVED;
+
+      let resolvedPrice = null;
+
+      // Fetch price from graph if available.
+      if (error) {
+        console.error(`Apollo client failed to fetch graph data:`, error);
+      } else if (hasPrice && !loading && data) {
+        // Implementation 1: Use subgraph data. This works for mainnet but not Kovan since the UMA subgraph does not currently index Kovan data.
```

Comment:
- I used the Kovan contracts `.getPrice()` calls to test on Kovan, but also verified that this query parsing works on mainnet using hardcoded, already resolved price request data.